src/main/java/edu/kis/powp/jobs2d/features/CommandFactory.java

```diff
@@ -0,0 +1,83 @@
+package edu.kis.powp.jobs2d.features;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import edu.kis.powp.jobs2d.command.CompoundCommand;
+import edu.kis.powp.jobs2d.command.DriverCommand;
+import edu.kis.powp.jobs2d.command.gui.CommandManagerWindow;
+import edu.kis.powp.jobs2d.command.manager.DriverCommandManager;
+import edu.kis.powp.observer.Subscriber;
+
+public class CommandFactory implements Subscriber {
+    public Map<String, DriverCommand> commandsMap = new HashMap<>();
+    private DriverCommandManager driverCommandManager;
+    private CommandManagerWindow window;
+
+    private static CommandFactory commandFactory = null;
+
+    public static CommandFactory getInstance() {
+        if (commandFactory == null) {
+            commandFactory = new CommandFactory();
+        }
+        return commandFactory;
+    }
+
+    private CommandFactory() {
+        driverCommandManager = CommandsFeature.getDriverCommandManager();
+        driverCommandManager.getChangePublisher().addSubscriber(this);
+    }
+
+
+    public DriverCommand getCommand(String commandName) throws CloneNotSupportedException {
+        DriverCommand resultCommand=null;
+        if(commandsMap.keySet().contains(commandName)){
+            resultCommand = commandsMap.get(commandName).clone();
+        }else{
+            throw new IllegalArgumentException("Name not found");
+        }
+
+        if (resultCommand instanceof CompoundCommand) {
+            ((CompoundCommand) resultCommand).setName(commandName);
+        }
```

Comment:
instanceof often points to a flaw in OO design. The cloning should be responsible for making a copy of a command name. Fix the problem, not the symptom.